src/main/java/software/amazon/event/kafkaconnector/EventBridgeSinkTask.java

```diff
@@ -59,102 +71,98 @@ public void put(Collection<SinkRecord> records) {
       return;
     }
 
-    var eventBridgeWriterRecords =
-        records.stream()
-            .map(this::convertToEventBridgeWriterRecord)
-            .filter(Objects::nonNull)
-            .collect(Collectors.toList());
+    var remainingRecords = List.copyOf(records);
 
-    // TODO: disabling batching for now until we can correctly calculate the size of the whole batch
-    // to avoid failing all entries in an EventBridge request due to exceeding 256kb size limit
-    var batches = Lists.partition(eventBridgeWriterRecords, 1);
-
-    var attempts = 0;
+    var attempts = new AtomicInteger(0);
     var start = OffsetDateTime.now(ZoneOffset.UTC);
-    while (true) {
-      attempts++;
+
+    while (!remainingRecords.isEmpty()) {
+      attempts.incrementAndGet();
       log.trace(
           "EventBridgeSinkTask putItems call started: start={} attempts={} maxRetries={}",
           start,
           attempts,
           config.maxRetries);
-      try {
-        batches.stream()
-            .flatMap(batch -> eventBridgeWriter.putItems(batch).stream())
-            .filter(EventBridgeWriterRecord::hasError)
-            .forEach(this::handleFailedEntries);
-        // only report successful putEvents calls
-        statusReporter.setSentRecords(eventBridgeWriterRecords.size());
-      } catch (EventBridgeWriterException e) {
-        switch (e.exceptionType) {
-          case RETRYABLE:
-            handleRetryableError(attempts, e);
-            continue; // retry
-          case NON_RETRYABLE:
-            handleNonRetryableError(e);
-        }
-      }
-      var completion = OffsetDateTime.now(ZoneOffset.UTC);
-      log.trace(
-          "EventBridgeSinkTask putItems call completed: start={} completion={} durationMillis={} "
-              + "attempts={} "
-              + "maxRetries={}",
-          start,
-          completion,
-          Duration.between(start, completion).toMillis(),
-          attempts,
-          config.maxRetries);
-      return;
-    }
-  }
 
-  private void handleNonRetryableError(EventBridgeWriterException e) {
-    log.error("Non-retryable failed put call: failing connector errorMessage={}", e.getMessage());
-    throw new ConnectException(e);
-  }
+      remainingRecords =
+          eventBridgeWriter.putItems(remainingRecords).stream()
+              .filter(EventBridgeResult::isFailure)
+              .flatMap(it -> handleFailedEntries(it.failure(), attempts.get()))
```

Comment:
Question: Do we have an issue with this retry/DLQ logic?
IIUC, any failed result is reported to DLQ but also retried until throwing?

Reply:
Nvm, I just saw that this is just the definition and the handling is in the writer as discussed 👍🏻

Wondering if this whole logic could be moved to the writer to simplify the task to a simple put call?

Reply:
Should be, but maybe in another issue/PR.